### ISSUE TITLE
Deferred input pipelines

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#15206]
Aside from the requests for a `$PSIndex`, index-variable in foreach pairing:
- [Enhance foreach loops to allow iterating over multiple, parallel collections in tandem (pairs) `#14732`](https://github.com/PowerShell/PowerShell/issues/14732)
- [new GPO operator .map `#15153`](https://github.com/PowerShell/PowerShell/issues/15153)
- [implement indexer in loop `#14724`](https://github.com/PowerShell/PowerShell/issues/14724)
- [Automatic variable for the pipeline index `#13772`](https://github.com/PowerShell/PowerShell/issues/13772)

It would also be nice if we could have parallel input **_streams_**

This:
```PowerShell
$Count = (1..3 |)
```
Causes currently an error:
> An empty pipe element is not allowed.

Instead it could possibly create a kind of a "**deferred pipeline object**" and each time the object is used/invoked it processes and returns the next item in the deferred pipeline (until it is empty where it returns an `AutomationNull`)

Wishful thinking:
```PowerShell
$Count = (1..3 |) # Initialize the deferred pipeline object
$Count
1
$Count
2
$Count
3
$Count # Nothing (`AutomationNull`) returns
```

More specific:
```PowerShell
$a = (Get-Connect .\MyHughFile.txt |) # or any other long stream
$b = (1..1e9 |) # In the idea, the range shouldn't affect the memory used for $b 
$c = (Import-Csv .\Large.csv |)
$a | ForEach-Object {
    Write-Host '$a item:' $_ 
    Write-Host '$b item:' $b # Every time the $b is used, it processes the next item in the deferred $b pipeline
    Write-Host '$c item:' $c # Ditto for $c
}
```

**Comment by vexx32** [comment]
I'm not sure this is that useful, to be honest. It reminds me a lot of `-PipelineVariable`, and I don't see a lot of folks really using that, either. It's a fairly complex thing to reason through and understand, iterating multiple pipelines at once.

Essentially what this would end up doing is conceptually and functionally similar to this (though, if it could be done as you describe, we'd want there to be less overhead involved).

```powershell
function next($enumerator) {
    if ($enumerator.MoveNext()) {
        $enumerator.Current
    }
}

$a = (1..1e9).GetEnumerator()

1..100 | ForEach-Object {
    Write-Host "Pipeline: $_"
    Write-Host "`$a item: $(next $a)"
}
```

I'm also not really sure that deferring the processing is going to be that valuable. For example, in your last example with $b, it's not possible to really defer the creation of the whole array without completely re-architecting how the `..` operator itself works. As soon as that script gets processed for the first time, the array has to be created in full, and then the enumerator captured so the later parts of the script can enumerate it.

Overall it's an interesting idea, but I'm concerned that it would be overly complex both to actually make use of as well as just to implement it. I think PowerShell's array creation and general pipeline semantics would need a bit of a rethink to make this a worthwhile thing to implement.

**Comment by iRon7** [comment]
@vexx32
>  It reminds me a lot of `-PipelineVariable`

I don't think it is the same as the `-PipelineVariable`. As the  `-PipelineVariable` assigns "current" items  (`$PSItem`) to different assignments in the **same** pipeline and the proposal concerns **separated** pipelines (and assignments). Although, I do agree that probably adds complexity.

> it's not possible to really defer the creation of the whole array without completely re-architecting how the `..` operator

I agree: it probably doesn't give any advantage for this specific statement (other than a possible syntactical advantage). Nevertheless, I wonder whether it could improve memory usage and performance¹ when I would load items from e.g. a file (`1..1e9 | Set-Content .\List.txt`). Unfortunately, I can't that this with your code example as the parentheses required for this '(Set-Content .\List.txt`).GetEnumerator()` would already load everything in memory...

<sub>**1.** in case I am actually not loading all lines as in the given example the `ForEach` will only loop a `100` times</sub>

**Comment by mklement0** [comment]
I like the idea too, but as suggested in https://github.com/PowerShell/PowerShell/issues/14732#issuecomment-817366599, perhaps it is sufficient to offer this object-at-a-time-in-lockstep processing via `ForEach-Object` rather than devising new syntax:

```powershell
# WISHFUL THINKING
{ 0..2 }, { 'a'..'c' } | ForEach-Object -InvokeScriptBlocks -Process { '{0}: {1}' -f $_[0], $_[1] }
0: a
1: b
2: c
```

That is, switch `-InvokeScriptBlocks` would interpret _script blocks_ passed as input as the input-data _generators_ to execute step by step, in tandem, via [`SteppablePipeline`](https://docs.microsoft.com/en-US/dotnet/api/System.Management.Automation.SteppablePipeline)s (as also used for [proxy functions](https://stackoverflow.com/a/62861781/45375)), with the collected results from each step passed to the `-Process` script block.

I tried to put together a quick proof-of-concept, but got stuck:

* I couldn't figure how to _capture_ the output from `.Process()` calls on steppable pipelines
* or how to detect when they're done producing output.

See my attempt below. @SeeminglyScience, any thoughts?

```powershell
function ForEach-SteppableObject {
  [CmdletBinding(PositionalBinding=$false)]
  param(
    [Parameter(Mandatory, Position=0)]
    [scriptblock] $Process
    ,
    [Parameter(Mandatory, ValueFromPipeline)]
    [scriptblock] $InputScriptBlock
  )

  begin {
    $allInputScriptBlocks = [System.Collections.Generic.List[scriptblock]]::new()
  }

  process {
    $allInputScriptBlocks.Add($InputScriptBlock)
  }

  end {

    [PowerShell[]] $psInstances = foreach ($sb in $allInputScriptBlocks) { [powershell]::Create().AddScript($sb) }
    
    [System.Management.Automation.SteppablePipeline[]] $steppablePipelines = $psInstances.GetSteppablePipeline()
  
    $steppablePipelines.Begin($false)

    do {
      # !! THIS IS THE PART THAT DOESN'T WORK.
      # !! The script blocks' output objects aren't being returned as part of the .Process()
      # !! call, and it's unclear to me how to get it to do that.
      [array] $outputs = $steppablePipelines.Process()

      # Pass the outputs from this step to the -Process script block.
      ForEach-Object -InputObject $outputs -Process $Process

    } while (($outputs -eq $null).Count -ne $outputs.Count) 
    # !! This currently results in an infinite loop, because each .Process()
    # !! call returns an empty array.
    # !! Without input being fed to the input script blocks, it's also not
    # !! clear to me how to reliably detect when a script block has stopped
    # !! producing.
  
    $steppablePipelines.End()

    $psInstances.Dispose()
  }

}

# Sample call:
# CAVEAT: Currently results in an infinite loop.
{ Write-Output (1..3) }, { Write-Output ('a'..'c') } | ForEach-SteppableObject { '{0}: {1}' -f $_[0], $_[1] } 
```

**Comment by iRon7** [comment]
Just a thought: there might be a performance advantage (for larger pipelines) if the deferred input pipeline runs really in parallel.

**Comment by SeeminglyScience** [comment]
Something like this is possible, and is probably the closest we can feasibly get:

<details>

<summary>PipelineEnumerable.cs</summary>

```csharp
using System;
using System.Collections;
using System.Management.Automation;

namespace Example
{
    public class PipelineEnumerable : IEnumerable
    {
        private readonly ScriptBlock _scriptBlock;

        private readonly IEnumerable _input;

        public PipelineEnumerable(ScriptBlock scriptBlock)
        {
            _scriptBlock = scriptBlock;
        }

        public PipelineEnumerable(ScriptBlock scriptBlock, IEnumerable input) {
            _scriptBlock = scriptBlock;
            _input = input;
        }

        public IEnumerator GetEnumerator()
        {
            return new PipelineEnumerator(
                _scriptBlock.GetSteppablePipeline(CommandOrigin.Runspace),
                _input?.GetEnumerator());
        }
    }

    public class PipelineEnumerator : IEnumerator, IDisposable
    {
        private readonly SteppablePipeline _pipe;

        private readonly IEnumerator _input;

        private IEnumerator _output;

        private bool _beginWasInvoked;

        private bool _processWasInvoked;

        private bool _endWasInvoked;

        private bool _isDisposed;

        public PipelineEnumerator(SteppablePipeline pipe, IEnumerator input)
        {
            _pipe = pipe;
            _input = input;
        }

        public object Current => _output.Current;

        public bool MoveNext()
        {
            if (!_beginWasInvoked)
            {
                bool expectingInput = _input is not null;
                _pipe.Begin(expectingInput);
                _beginWasInvoked = true;
                // Always forget `SteppablePipeline.Begin` has a void return for some reason.
                // if (($this.out = ($res)?.GetEnumerator())?.MoveNext()) {
                //     return $true
                // }
            }

            if (_output?.MoveNext() is true)
            {
                return true;
            }

            while (_input?.MoveNext() is true)
            {
                Array res = _pipe.Process(_input.Current);
                _processWasInvoked = true;
                if ((_output = res?.GetEnumerator())?.MoveNext() is true)
                {
                    return true;
                }
            }

            if (!_processWasInvoked)
            {
                Array res = _pipe.Process();
                _processWasInvoked = true;
                if ((_output = res?.GetEnumerator())?.MoveNext() is true)
                {
                    return true;
                }
            }

            if (!_endWasInvoked)
            {
                Array res = _pipe.End();
                _endWasInvoked = true;
                if ((_output = res?.GetEnumerator())?.MoveNext() is true)
                {
                    return true;
                }
            }

            return false;
        }

        public void Reset() => throw new NotSupportedException();

        protected virtual void Dispose(bool disposing)
        {
            if (_isDisposed)
            {
                return;
            }

            if (disposing)
            {
                _pipe.Dispose();
            }

            _isDisposed = true;
        }

        public void Dispose()
        {
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }
    }
}
```

</details>

This sort of highlights the problem with the idea though.  There isn't a way to defer execution after a single pipeline object has been emitted.

For example, if you were to do this:

```powershell
$b = [Example.PipelineEnumerable]::new{
	ForEach-Object {
		foreach ($a in 0..3) {
			$a
			$host.UI.WriteErrorLine('ehhh')
		}
	}
}

$e = $b.GetEnumerator()
$e.MoveNext()
$e.Current
```

You get

```raw
ehhh
ehhh
ehhh
ehhh
True
0
```

During a pipeline this works because when one item is emitted to the pipeline, execution moves immediately to the next pipeline element's process block.  Without an immediate next method to call, you **must** finish the block. There is no way to suspend and resume execution of a method in dotnet.  Even when C# *seems* to do this (e.g. with async/await and `yield return`), the compiler is really completely changing the method to enable this.

**Comment by mklement0** [comment]
Great stuff, @SeeminglyScience; thanks for clarifying the limitations of steppable pipelines outside of actual pipelines.

I played around with parallel thread jobs as a - suboptimal - alternative: the `ForEach-CorrespondingObject` function defined below uses _queues_ to buffer output objects in order to ensure that only corresponding output objects are processed; with large output sets and widely differing output timing among the input script blocks, however, you may still run out of memory.

Of course, this approach isn't efficient, because buffering happens both in the thread jobs and in the function, and output is periodically collected with a short sleep interval between tries.

Here's a sample call and its output, with verbose output enabled:

* Two infinite producer script blocks are supplied, which simply output as sequence of numbers, and a the processing script block simple echoes the corresponding number pairs.
  * One of the blocks periodically cleans the screen, to avoid memory issues from overly long console scrollback buffers.
  
* Note that the producer do not run in a _tight_ loop - they sleep briefly between outputting numbers - as there is currently no batching mechanism for `Receive-Object` calls, which could cause a given call to receive objects indefinitely; see:
   * #15234 - the inability to retrieve a given number of objects from a job in a consuming manner.
   * #5433. - proposal to call `Receive-Object` with a timeout.

```powershell
{ 
  $i=0; $sleepMs = 10
  while ($true) { 
    (++$i); Start-Sleep -ms $sleepMs
  } 
}, 
{ 
  $i=0; $sleepMs = 10
  while ($true) { 
    (++$i); Start-Sleep -ms $sleepMs
    # To avoid memory problems with an overly long scrollback buffer, we clear it
    # periodically.
  if ($i -gt 1000 -and $i % 1000 -eq 1) { (-not $IsWindows -or [bool] $env:WT_SESSION) ? [Console]::Write("`e[2J`e[3J`e[H") : [Console]::Clear(); [Console]::WriteLine('(Screen cleared.)') } 
  } 
} | ForEach-CorrespondingObject { '{0} - {1}' -f $_[0], $_[1] } -Verbose
```

Sample output:

![image](https://user-images.githubusercontent.com/588825/114780602-a0bd5900-9d45-11eb-94e8-b2344d6bd22c.png)


<details>
<summary><pre>ForEach-CorrespondingObject</pre> source code.</summary>

```powershell
# Sample call.
function ForEach-CorrespondingObject {
  [CmdletBinding(PositionalBinding = $false)]
  param(
    [Parameter(Mandatory, Position = 0)]
    [scriptblock] $Process
    ,
    [Parameter(Mandatory, ValueFromPipeline)]
    [scriptblock] $InputScriptBlock
  )
  begin {
    $allInputScriptBlocks = [System.Collections.Generic.List[scriptblock]]::new()
  }
  process {
    $allInputScriptBlocks.Add($InputScriptBlock)
  }
  end {
    # Start a thread job for each script block.
    Write-Verbose "Launching $($allInputScriptBlocks.Count) jobs..."
    [array] $jobs = foreach ($sb in $allInputScriptBlocks) {
      Start-ThreadJob -ThrottleLimit $allInputScriptBlocks.Count $sb
    }
    try {
      # Initialize an array of output queues parallelling the jobs and
      # temporarily storing their output objects.
      $outputQueues = [Collections.Queue[]]::new($jobs.Count)
      for ($i = 0; $i -lt $outputQueues.Count; ++$i) { $outputQueues[$i] = [Collections.Queue]::new() }
      # Start collecting output.
      do {
        Write-Verbose "(Re)entering output-collection loop..."
        # Sleep a little between attempts to collect output.
        Start-Sleep -ms 50
        # Collect currently available output.
        for ($i = 0; $i -lt $jobs.Count; ++$i) {
          $job = $jobs[$i]
          if (-not $job) {
            # Set a completed job's output queue to $null, if it is now empty.
            if ($outputQueues[$i] -and $outputQueues[$i].Count -eq 0) { $outputQueues[$i] = $null }
          }
          else {
            $done = $job.State -in 'Completed', 'Failed'
            if ($job.HasMoreData) {
              Write-Verbose "Receiving from job #$i..."
              # Note: using Select-Object -First <n> is unfortunately NOT an option: see https://github.com/PowerShell/PowerShell/issues/15234
              # With a tight loop as input, this can end up collecting a lot of data.
              Receive-Job -Job $job | ForEach-Object {
                $outputQueues[$i].Enqueue($_)
              }
            }
            else {
              Write-Verbose "Job #$i has no data."
            }
            if (-not $job.HasMoreData -and $done) {
              $jobs[$i] = $null; Remove-Job $job
            }
          }
        }
    
        # Process all corresponding objects currently in the queues.
        if ($activeQueues = $outputQueues -ne $null) {
          $availableOutputSetsCount = [Linq.Enumerable]::Min([int[]] $activeQueues.ForEach('Count'))
          Write-Verbose "Processing $($availableOutputSetsCount) available output set(s)..."
          for ($i = 0; $i -lt $availableOutputSetsCount; ++$i) {
            [array] $outputSet = foreach ($queue in $outputQueues) { $queue -and $queue.Count ? $queue.Dequeue() : $null }
            # Pass each set of corresponding output objects to the -Process script block.
            ForEach-Object -InputObject $outputSet -Process $Process
          }
        }
      } while ($jobs -ne $null -or $activeQueues)
    }
    finally {
      if ($activeJobs = $jobs -ne $null) {
        Remove-Job -Job $activeJobs -Force
      }
    }
  }
}
```

</details>

**Comment by iRon7** [comment]
Thanks for the commends and prototypes.

@SeeminglyScience,
Unfortuanetely, I can get you C example to run as I get an error:

> ```
> Line |
>  132 |  Add-Type -TypeDefinition $Source
>      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>      | (59,49): error CS8652: The feature 'not pattern' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.                 bool expectingInput =
>      | _input is not null;                                                 ^
> ```

The error is quiet clear but I have you clue how to resolve this (and is I want this, maybe I should do some more playing around in a sandbox).

@mklement0,
I have had it some deeper thoughts and I agree that this shouldn't require a different syntax.
But I think that the `ForEach-Object` should have an additional parameter instead, something like `-SynchronicallyProcess <ScriptBlock[]>` knowing that the parallel pipelines supplied are actually not a part of the (main) pipeline but (initially) nothing more than static scriptblocks.

What I have for testing:
```
$Command = { GCI }                                              # a Synchronic Process
$Sync = [hashtable]::Synchronized(@{
    Command = $Command
    WaitMain = [System.Threading.AutoResetEvent]::new($True)
    WaitJob = [System.Threading.AutoResetEvent]::new($False)
    Output = $Null
})

$Script = {
    param($Sync)
    Invoke-Command $Sync['Command'] |
    ForEach-Object { 
        $Sync['Output'] = $_                                    # Should be a data collector?
        $Null = $Sync['WaitJob'].Set()                          # For multiple parallel pipelines use a barrier?
        $Null = $Sync['WaitMain'].WaitOne(10000)
    }
    $Sync['Output'] = &{}
    $Null = $Sync['WaitJob'].Set()
}

$PowerShell = [PowerShell]::Create()
$Null = $PowerShell.AddScript($Script).AddArgument($Sync)
$Job = $PowerShell.BeginInvoke()
Sleep -Milliseconds 500                                         #7626 Prevent a "Stack Empty"
1..9 |ForEach-Object {                                          # Start the main pipeline
    $SyncItem = if ($Null -ne $PowerShell) {
        $Null = $Sync['WaitMain'].Set()                         # For multiple parallel pipelines use a barrier?
        $Null = $Sync['WaitJob'].WaitOne(5000)
        $Output = $Sync['Output']                               # $powerShell.Invoke ?
        If ($Null -ne $Output) {
            $Output
        }
        else {
            $PowerShell.Dispose()
            $PowerShell = $Null
        }
    }
    '{0} - {1}' -f $_, $SyncItem
}
```

Unfortunately, the result is quite slow, besides I am lacking detailed knowledge of thread signaling, data collectors and I running into bugs (#7626) to make this stable.
Anyways, for the highlevel design, I think the way to implement this is would be:
- Initiate a thread for each scripblock (supplied by the proposed `-SynchronicallyProcess ` parameter)
- in the thread, `Invoke-Command` the scriptblock and pipe it to a `ForEach-Object`
- For each item, share it with the main script (using a [`PSDataCollector`](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.psdatacollection-1)?)
- Block the pipeline in the script (with a [`Barrier`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.barrier?view=net-5.0)?) ***after*** each item
- Take care of the fact that the parallel thread might run out of objects.

In the main thread:
- (Initiate a thread for each scripblock (supplied by the proposed `-SynchronicallyProcess ` parameter))
- Start the main pipeline (`ForEach-Object`)
- ***Before*** processing each current item (`$_`), implement a barrier (shared with all the above process)
to wait for all the parallel thread to supply their next item.
- Make the parallel items available (e.g. in a `$SyncItem[]`) and process the current

**Comment by SeeminglyScience** [comment]
> The error is quiet clear but I have you clue how to resolve this (and is I want this, maybe I should do some more playing around in a sandbox).

You can update PowerShell or alternatively just change any instances of that pattern (e.g. `is null` to `== null` and `is not null` to `!= null`)

---

re: the thread based version.  If every deferred pipeline spawns a new thread that could get out of hand pretty quickly if folks don't know how heavy it is.  On top of that you'll end up losing state since you'll also need to spawn a new runspace for each one, which I don't think the user will expect for this functionality.

It's a cool idea, but I'm not sure it solves enough problems to justify the overhead, usage complication or implementation challenge.

**Comment by mklement0** [comment]
@SeeminglyScience:

Good point about not being able to use the caller's state being a limitation that may surprise users, but the same applies at least in principle to `ForEach -Parallel`, where you need `$using:` to reference the caller's state. How would/can something like `$using` work with a thread-based approach?

I'd expect the number of input blocks to be limited, so perhaps the thread count is even less of a concern than with `ForEach-Object -Parallel`.

@iRon7:

> I have had it some deeper thoughts and I agree that this shouldn't require a different syntax.
But I think that the ForEach-Object should have an additional parameter instead,

Yes, that was [my suggestion too](https://github.com/PowerShell/PowerShell/issues/15206#issuecomment-818269432) (I only used a different name for the proof-of-concept, because it wasn't a superset of `ForEach-Object`).
I've been struggling to come up with a good switch name, however, given that it would ideally convey _two_ aspects: accepting _script blocks_ as input, and processing their output in a synchronized fashion.

**Comment by SeeminglyScience** [comment]
> @SeeminglyScience:
> 
> Good point about not being able to use the caller's state being a limitation that may surprise users, but the same applies at least in principle to `ForEach -Parallel`, where you need `$using:` to reference the caller's state. How would/can something like `$using` work with a thread-based approach?

I imagine it could be used in much the same way.  This would most likely need to be implemented as a custom sort of `Out-Default`, not much different than `-Parallel` otherwise. The hard part is communicating to the user that they need it for this, and ensuring that it's worth the trouble.

**Comment by rjmholt** [comment]
We discussed this in the Engine Working Group today and while this is a compelling idea, we think:

- The proposed functionality doesn't mesh well with PowerShell's existing enumeration paradigm
- Other alternatives exist for implementing lazy enumeration in PowerShell today
- There's not a clear justification for needing to first-class a mechanism like this at a language/syntax level
- The proposed pipeline syntax itself would have consequences for parsing pipelines, especially interactively
- The number and extent of changes that would be required to implement this would be difficult to justify compared to the benefit of the feature
- It's not clear that there's a lot of demand or potential utility for this proposal
- Without an example implementation, a proposal of this size requires a clear demand or scenario to motivate it

With that in mind, we are declining the feature request.
